version: '3'

services:
  # Attackers
  session_dos:
    build: ./cpp
    volumes:
      - ./cpp:/code
    command: actors/session_dos
    networks:
      testing_net:
        ipv4_address: 172.28.1.5
    depends_on:
      - generic_scanner

  io_dos:
    build: ./cpp
    volumes:
      - ./cpp:/code
    command: actors/io_dos
    networks:
      testing_net:
        ipv4_address: 172.28.1.5
    depends_on:
      - io_scanner

  io_dos_debug:
    # Used to read privilege error
    build: ./python
    command: python /usr/src/app/mitm.py 172.28.1.1 172.28.1.3
    user: root
    networks:
      testing_net:
        ipv4_address: 172.28.1.5
    depends_on:
      - io_dos
    sysctls:
      - net.ipv4.ip_forward=1
      - net.ipv6.conf.all.forwarding=0
      - net.ipv4.conf.all.send_redirects=0

  mitm:
    build: ./python
    command: python /usr/src/app/mitm.py 172.28.1.1 172.28.1.3
    user: root
    networks:
      testing_net:
        ipv4_address: 172.28.1.5
    depends_on:
      - generic_scanner
    sysctls:
      # Necessary for transparent mode for mitmproxy
      - net.ipv4.ip_forward=1
      - net.ipv6.conf.all.forwarding=0
      - net.ipv4.conf.all.send_redirects=0

  # Legitimate services
  generic_scanner:
    build: ./cpp
    volumes:
      - ./cpp:/code
    command: actors/generic_scanner
    networks:
      testing_net:
        ipv4_address: 172.28.1.1
    depends_on:
      - opener_adapter

  io_scanner:
    build: ./cpp
    volumes:
      - ./cpp:/code
    command: actors/io_scanner
    networks:
      testing_net:
        ipv4_address: 172.28.1.1
    depends_on:
      - opener_adapter

  opener_adapter:
    build: ./cpp/EIPScanner/docker/opener
    working_dir: /OpENer/bin/posix/src/ports/POSIX
    command: ./OpENer eth0
    ports:
      - "44818:44818"
    networks:
      testing_net:
        ipv4_address: 172.28.1.3
networks:
  testing_net:
    ipam:
      driver: default
      config:
        - subnet: 172.28.0.0/16
